### ISSUE TITLE
Login succeed then redirect, redirect, redirect... Finally page isn't redirecting properly.

### ISSUE THREAD

**Issue by relf** [jupyterhub/jupyterhub-deploy-docker#34]
Hello,
As I'am trying to install jupyterhub using docker on the same host I've started from your configuration and just modified the authenticator to be LDAPAuthenticator.
After login, I get the 'Start Server' page but after clicking on the start button firefox displays 'the page isn't redirecting properly'.

Here the logs of the jupyterhub:
<pre>
Starting jupyterhub ...
Starting jupyterhub ... done
Attaching to jupyterhub
jupyterhub | [I 2017-05-16 08:08:17.071 JupyterHub app:750] Loading cookie_secret from /data/jupyterhub_cookie_secret
jupyterhub | [W 2017-05-16 08:08:17.144 JupyterHub app:376]
jupyterhub |     Generating CONFIGPROXY_AUTH_TOKEN. Restarting the Hub will require restarting the proxy.
jupyterhub |     Set CONFIGPROXY_AUTH_TOKEN env or JupyterHub.proxy_auth_token config to avoid this message.
jupyterhub |
jupyterhub | [W 2017-05-16 08:08:17.156 JupyterHub app:908] No admin users, admin interface will be unavailable.
jupyterhub | [W 2017-05-16 08:08:17.156 JupyterHub app:909] Add any administrative users to `c.Authenticator.admin_users` in config.
jupyterhub | [I 2017-05-16 08:08:17.156 JupyterHub app:936] Not using whitelist. Any authenticated user will be allowed.
jupyterhub | [I 2017-05-16 08:08:17.190 JupyterHub dockerspawner:360] Container 'jupyter-rlafage' is gone
jupyterhub | [W 2017-05-16 08:08:17.191 JupyterHub dockerspawner:331] container not found
jupyterhub | [I 2017-05-16 08:08:17.212 JupyterHub app:1536] Hub API listening on http://jupyterhub:8080/hub/
jupyterhub | [I 2017-05-16 08:08:17.219 JupyterHub app:1255] Starting proxy @ http://*:443/
jupyterhub | 08:08:17.556 - info: [ConfigProxy] Proxying https://*:443 to http://jupyterhub:8080
jupyterhub | 08:08:17.561 - info: [ConfigProxy] Proxy API at http://127.0.0.1:444/api/routes
jupyterhub | [I 2017-05-16 08:08:17.631 JupyterHub app:1592] JupyterHub is now running at http://127.0.0.1:443/
jupyterhub | [I 2017-05-16 08:08:25.977 JupyterHub log:100] 302 GET / (@::ffff:134.212.26.254) 2.09ms
jupyterhub | [I 2017-05-16 08:08:26.160 JupyterHub log:100] 302 GET /hub (@::ffff:134.212.26.254) 0.76ms
jupyterhub | [I 2017-05-16 08:08:26.190 JupyterHub log:100] 302 GET /hub/ (rlafage@::ffff:134.212.26.254) 5.60ms
jupyterhub | [I 2017-05-16 08:08:26.251 JupyterHub log:100] 200 GET /hub/home (rlafage@::ffff:134.212.26.254) 52.45ms
jupyterhub | [I 2017-05-16 08:08:31.128 JupyterHub log:100] 302 GET /spawn?next=%2Fhub%2Fhome (@::ffff:134.212.26.254) 2.98ms
jupyterhub | [I 2017-05-16 08:08:31.148 JupyterHub log:100] 302 GET /hub/spawn?next=%2Fhub%2Fhome (rlafage@::ffff:134.212.26.254) 6.86ms
jupyterhub | [I 2017-05-16 08:08:31.155 JupyterHub log:100] 302 GET /user/rlafage (@::ffff:134.212.26.254) 1.32ms
jupyterhub | [I 2017-05-16 08:08:31.173 JupyterHub dockerspawner:360] Container 'jupyter-rlafage' is gone
jupyterhub | [W 2017-05-16 08:08:31.173 JupyterHub dockerspawner:331] container not found
jupyterhub | [I 2017-05-16 08:08:31.584 JupyterHub dockerspawner:360] Container 'jupyter-rlafage' is gone
jupyterhub | [I 2017-05-16 08:08:31.819 JupyterHub dockerspawner:433] Created container 'jupyter-rlafage' (id: 4d09164) from image jupyterhub/singleuser:latest
jupyterhub | [I 2017-05-16 08:08:31.819 JupyterHub dockerspawner:450] Starting container 'jupyter-rlafage' (id: 4d09164)
jupyterhub | [I 2017-05-16 08:08:34.590 JupyterHub base:341] User rlafage server took 3.417 seconds to start
jupyterhub | [I 2017-05-16 08:08:34.590 JupyterHub orm:198] Adding user rlafage to proxy /user/rlafage => http://172.18.0.3:8888
jupyterhub | [I 2017-05-16 08:08:34.617 JupyterHub log:100] 302 GET /hub/user/rlafage (rlafage@::ffff:134.212.26.254) 3448.68ms
jupyterhub | [W 2017-05-16 08:08:34.649 JupyterHub pages:42] Redirecting /hub/?next=%2Fuser%2Frlafage%2Ftree%3F to /hub/user/rlafage/tree?. For sharing public links, use /user-redirect/
jupyterhub | [I 2017-05-16 08:08:34.657 JupyterHub log:100] 302 GET /hub/?next=%2Fuser%2Frlafage%2Ftree%3F (rlafage@::ffff:134.212.26.254) 2.14ms
jupyterhub | [I 2017-05-16 08:08:34.684 JupyterHub log:100] 302 GET /hub/user/rlafage/tree? (rlafage@::ffff:134.212.26.254) 15.89ms
jupyterhub | [W 2017-05-16 08:08:34.692 JupyterHub pages:42] Redirecting /hub/?next=%2Fuser%2Frlafage%2Ftree%3F to /hub/user/rlafage/tree?. For sharing public links, use /user-redirect/
jupyterhub | [I 2017-05-16 08:08:34.701 JupyterHub log:100] 302 GET /hub/?next=%2Fuser%2Frlafage%2Ftree%3F (rlafage@::ffff:134.212.26.254) 2.05ms
jupyterhub | [I 2017-05-16 08:08:34.730 JupyterHub log:100] 302 GET /hub/user/rlafage/tree? (rlafage@::ffff:134.212.26.254) 15.56ms
jupyterhub | [W 2017-05-16 08:08:34.735 JupyterHub pages:42] Redirecting /hub/?next=%2Fuser%2Frlafage%2Ftree%3F to /hub/user/rlafage/tree?. For sharing public links, use /user-redirect/
jupyterhub | [I 2017-05-16 08:08:34.743 JupyterHub log:100] 302 GET /hub/?next=%2Fuser%2Frlafage%2Ftree%3F (rlafage@::ffff:134.212.26.254) 1.74ms
jupyterhub | [I 2017-05-16 08:08:34.772 JupyterHub log:100] 302 GET /hub/user/rlafage/tree? (rlafage@::ffff:134.212.26.254) 15.15ms
jupyterhub | [W 2017-05-16 08:08:34.793 JupyterHub pages:42] Redirecting /hub/?next=%2Fuser%2Frlafage%2Ftree%3F to /hub/user/rlafage/tree?. For sharing public links, use /user-redirect/
jupyterhub | [I 2017-05-16 08:08:34.802 JupyterHub log:100] 302 GET /hub/?next=%2Fuser%2Frlafage%2Ftree%3F (rlafage@::ffff:134.212.26.254) 1.90ms
jupyterhub | [I 2017-05-16 08:08:34.828 JupyterHub log:100] 302 GET /hub/user/rlafage/tree? (rlafage@::ffff:134.212.26.254) 15.32ms
jupyterhub | [W 2017-05-16 08:08:34.838 JupyterHub pages:42] Redirecting /hub/?next=%2Fuser%2Frlafage%2Ftree%3F to /hub/user/rlafage/tree?. For sharing public links, use /user-redirect/
jupyterhub | [I 2017-05-16 08:08:34.847 JupyterHub log:100] 302 GET /hub/?next=%2Fuser%2Frlafage%2Ftree%3F (rlafage@::ffff:134.212.26.254) 2.04ms
jupyterhub | [I 2017-05-16 08:08:34.874 JupyterHub log:100] 302 GET /hub/user/rlafage/tree? (rlafage@::ffff:134.212.26.254) 16.10ms
</pre>

and here the logs of the jupyter/singleuser:
<pre>
[rlafage@rdri206h jupyterhub-deploy-docker] docker logs jupyter-rlafage
[W 2017-05-16 08:08:34.398 rlafage configurable:168] Config option `open_browser` not recognized by `SingleUserNotebookApp`.  Did you mean `browser`?
[I 2017-05-16 08:08:34.411 rlafage notebookapp:592] Writing notebook server cookie secret to /home/jovyan/.local/share/jupyter/runtime/notebook_cookie_secret
[W 2017-05-16 08:08:34.500 rlafage login:225] All authentication is disabled.  Anyone who can connect to this server will be able to run code.
[I 2017-05-16 08:08:34.522 rlafage extension:193] JupyterLab alpha preview extension loaded from /opt/conda/lib/python3.5/site-packages/jupyterlab
[I 2017-05-16 08:08:34.533 rlafage notebookapp:1366] Serving notebooks from local directory: /home/jovyan/work
[I 2017-05-16 08:08:34.534 rlafage notebookapp:1366] 0 active kernels
[I 2017-05-16 08:08:34.534 rlafage notebookapp:1366] The Jupyter Notebook is running at: http://0.0.0.0:8888/user/rlafage/
[I 2017-05-16 08:08:34.534 rlafage notebookapp:1367] Use Control-C to stop this server and shut down all kernels (twice to skip confirmation).
[I 2017-05-16 08:08:34.588 rlafage log:47] 302 GET /user/rlafage (172.18.0.2) 1.22ms
[I 2017-05-16 08:08:34.622 rlafage log:47] 302 GET /user/rlafage (::ffff:134.212.26.254) 0.71ms
[I 2017-05-16 08:08:34.639 rlafage log:47] 302 GET /user/rlafage/tree? (::ffff:134.212.26.254) 1.08ms
[I 2017-05-16 08:08:34.685 rlafage log:47] 302 GET /user/rlafage/tree? (::ffff:134.212.26.254) 1.11ms
[I 2017-05-16 08:08:34.725 rlafage log:47] 302 GET /user/rlafage/tree? (::ffff:134.212.26.254) 1.11ms
[I 2017-05-16 08:08:34.769 rlafage log:47] 302 GET /user/rlafage/tree? (::ffff:134.212.26.254) 1.14ms
[I 2017-05-16 08:08:34.829 rlafage log:47] 302 GET /user/rlafage/tree? (::ffff:134.212.26.254) 1.35ms
[I 2017-05-16 08:08:34.870 rlafage log:47] 302 GET /user/rlafage/tree? (::ffff:134.212.26.254) 1.19ms
</pre>

I found that related issue [jupyterhub #742](https://github.com/jupyterhub/jupyterhub/issues/742) and tried various configuration changes without success.

Please, could you give me a hint? What could I check to make it through this?
Thanks in advance.

**Comment by relf** [comment]
... and to complete below my configuration files. 
Main changes to be noticed: 
* DOCKER_NOTEBOOK_IMAGE=jupyterhub/singleuser:latest
* LdapAuthenticator
* c.DockerSpawner.volumes = {'/d/rlafage' : notebook_dir }

<pre>
# values for environment variables in the configuration file IF the variables
# are not set in the shell environment.

# To override these values, set the shell environment variables.

# Name of Docker machine
export DOCKER_MACHINE_NAME=jupyterhub

# Name of Docker network
export DOCKER_NETWORK_NAME=jupyterhub-network

# Single-user Jupyter Notebook server container image
export DOCKER_NOTEBOOK_IMAGE=jupyterhub/singleuser:latest

# Notebook directory in the container.
# This will be /home/jovyan/work if the default
# This directory is stored as a docker volume for each user
export DOCKER_NOTEBOOK_DIR=/home/jovyan/work

# Docker run command to use when spawning single-user containers
export DOCKER_SPAWN_CMD=start-singleuser.sh

# Name of JupyterHub container data volume
export DATA_VOLUME_HOST=jupyterhub-data

# Data volume container mount point
export DATA_VOLUME_CONTAINER=/data
</pre>

Dockerfile.jupyterhub
<pre>
# Copyright (c) Jupyter Development Team.
# Distributed under the terms of the Modified BSD License.
FROM jupyterhub/jupyterhub

ENV HTTP_PROXY=proxy.onecert.fr:80
ENV HTTPS_PROXY=proxy.onecert.fr:80

# Install dockerspawner and its dependencies
RUN /opt/conda/bin/pip install jupyterhub-ldapauthenticator 

RUN git clone https://github.com/jupyterhub/dockerspawner.git
WORKDIR dockerspawner
RUN pip install -r requirements.txt
RUN pip install .
WORKDIR ..

# Copy TLS certificate and key
ENV SSL_CERT /srv/jupyterhub/secrets/jupyterhub.pem
ENV SSL_KEY /srv/jupyterhub/secrets/jupyterhub.key
COPY ./secrets/*.pem $SSL_CERT
COPY ./secrets/*.key $SSL_KEY
RUN chmod 700 /srv/jupyterhub/secrets && \
    chmod 600 /srv/jupyterhub/secrets/*

CMD ["jupyterhub"]
</pre>

jupyter_config.py
<pre>
# Copyright (c) Jupyter Development Team.
# Distributed under the terms of the Modified BSD License.

# Configuration file for JupyterHub
import os

c = get_config()

# We rely on environment variables to configure JupyterHub so that we
# avoid having to rebuild the JupyterHub container every time we change a
# configuration parameter.

# Spawn single-user servers as Docker containers
c.JupyterHub.spawner_class = 'dockerspawner.DockerSpawner'
# Spawn containers from this image
c.DockerSpawner.container_image = os.environ['DOCKER_NOTEBOOK_IMAGE']
# JupyterHub requires a single-user instance of the Notebook server, so we
# default to using the `start-singleuser.sh` script included in the
# jupyter/docker-stacks *-notebook images as the Docker run command when
# spawning containers.  Optionally, you can override the Docker run command
# using the DOCKER_SPAWN_CMD environment variable.
spawn_cmd = os.environ.get('DOCKER_SPAWN_CMD', "start-singleuser.sh")
c.DockerSpawner.extra_create_kwargs.update({ 'command': spawn_cmd })
# Connect containers to this Docker network
network_name = os.environ['DOCKER_NETWORK_NAME']
c.DockerSpawner.use_internal_ip = True
c.DockerSpawner.network_name = network_name
# Pass the network name as argument to spawned containers
c.DockerSpawner.extra_host_config = { 'network_mode': network_name }
# Explicitly set notebook directory because we'll be mounting a host volume to
# it.  Most jupyter/docker-stacks *-notebook images run the Notebook server as
# user `jovyan`, and set the notebook directory to `/home/jovyan/work`.
# We follow the same convention.
notebook_dir = os.environ.get('DOCKER_NOTEBOOK_DIR') or '/home/jovyan/work'
c.DockerSpawner.notebook_dir = notebook_dir
# Mount the real user's Docker volume on the host to the notebook user's
# notebook directory in the container
c.DockerSpawner.volumes = {'/d/rlafage' : notebook_dir }
c.DockerSpawner.extra_create_kwargs.update({ 'volume_driver': 'local' })
# Remove containers once they are stopped
c.DockerSpawner.remove_containers = True
# For debugging arguments passed to spawned containers
c.DockerSpawner.debug = True

# User containers will access hub by container name on the Docker network
c.JupyterHub.hub_ip = 'jupyterhub'
c.JupyterHub.hub_port = 8080

# TLS config
c.JupyterHub.port = 443
c.JupyterHub.ssl_key = os.environ['SSL_KEY']
c.JupyterHub.ssl_cert = os.environ['SSL_CERT']

# Authenticate users with GitHub OAuth
#c.JupyterHub.authenticator_class = 'oauthenticator.GitHubOAuthenticator'
#c.GitHubOAuthenticator.oauth_callback_url = os.environ['OAUTH_CALLBACK_URL']
c.JupyterHub.authenticator_class = 'ldapauthenticator.LDAPAuthenticator'
c.LDAPAuthenticator.server_address = 'ldap'
c.LDAPAuthenticator.bind_dn_template = 'uid={username},ou=people,o=onera'
c.LDAPAuthenticator.use_ssl = True

# Persist hub data on volume mounted inside container
data_dir = os.environ.get('DATA_VOLUME_CONTAINER', '/data')
c.JupyterHub.db_url = os.path.join('sqlite:///', data_dir, 'jupyterhub.sqlite')
c.JupyterHub.cookie_secret_file = os.path.join(data_dir,
    'jupyterhub_cookie_secret')

# Whitlelist users and admins
c.Authenticator.whitelist = whitelist = set()
c.Authenticator.admin_users = admin = set()
c.JupyterHub.admin_access = True
#pwd = os.path.dirname(__file__)
#with open(os.path.join(pwd, 'userlist')) as f:
#    for line in f:
#        if not line:
#            continue
#        parts = line.split()
#        name = parts[0]
#        whitelist.add(name)
#        if len(parts) > 1 and parts[1] == 'admin':
#            admin.add(name)
</pre>

docker-compose.yml
<pre>
services:
  hub:
    build:
      context: .
      dockerfile: Dockerfile.jupyterhub
    image: jupyterhub
    container_name: jupyterhub
    volumes:
      - "/d/rlafage/rdri206h/jupyterhub-deploy-docker:/srv/jupyterhub"
      # Bind Docker socket on the host so we can connect to the daemon from
      # within the container
      - "/var/run/docker.sock:/var/run/docker.sock:rw"
      # Bind Docker volume on host for JupyterHub database and cookie secrets
      - "data:${DATA_VOLUME_CONTAINER}"
    ports:
      - "443:443"
    environment:
      # All containers will join this network
      DOCKER_NETWORK_NAME: ${DOCKER_NETWORK_NAME}
      # JupyterHub will spawn this Notebook image for users
      DOCKER_NOTEBOOK_IMAGE: ${DOCKER_NOTEBOOK_IMAGE}
      # Notebook directory inside user image
      DOCKER_NOTEBOOK_DIR: ${DOCKER_NOTEBOOK_DIR}
      # Using this run command (optional)
      DOCKER_SPAWN_CMD: ${DOCKER_SPAWN_CMD}
      # Required to authenticate users using GitHub OAuth
      #GITHUB_CLIENT_ID: ${GITHUB_CLIENT_ID}
      #GITHUB_CLIENT_SECRET: ${GITHUB_CLIENT_SECRET}
      #OAUTH_CALLBACK_URL: ${OAUTH_CALLBACK_URL}
    command: >
      jupyterhub -f /srv/jupyterhub/jupyterhub_config.py

volumes:
  data:
    external:
      name: ${DATA_VOLUME_HOST}

networks:
  default:
    external:
      name: ${DOCKER_NETWORK_NAME}
</pre>

**Comment by smartquant** [comment]
Hello,
I had the very same issue as @relf. What solved it for me is to still use version 0.7 of jupyterhub and using dockerspawner 0.5, as is the case in the Dockerfile.jupyterhub build file provided in this repo. 

```
FROM jupyterhub/jupyterhub-onbuild:0.7.0

# Install dockerspawner and its dependencies
RUN /opt/conda/bin/pip install \
    dockerspawner==0.5.* \
    jupyterhub-ldapauthenticator
```

**Comment by jtyberg** [comment]
Thank you for the feedback @smartquant. Which version of `dockerspawner` were you using when you experienced the redirect issue? 

@relf Are you still experiencing this issue? And what version of `jupyterhub` and `dockerspawner` are you using?

As this repo is intended to be a reference deployment of JHub on Docker, I only test/update on occasion, when new versions of dependencies are released. I suspect there may have been a change in `dockerspawner` that requires an update to this deployment.

**Comment by relf** [comment]
I ended up by installing jupyterhub 0.7.2 directly without docker and it works.

**Comment by jtyberg** [comment]
I think a lot of fixes have gone in since JupyterHub 0.7.0. I just bumped to version 0.7.2, and `dockerspawner` to version 0.7.0 in this repo (#35) and tested locally. Everything came up fine for me.

**Comment by willingc** [comment]
Closed by #35

**Comment by pgeez** [comment]
running Jupyterhub 0.7.2 in a container with the DockerSpawner works, but running Jupyterhub 0.8.0 (latest on Dockerhub) causes the same infinite redirect issue and does not work.

https://github.com/jupyterhub/jupyterhub/issues/709
